### PR TITLE
 Support fe heartbeat use thrift protocol to get stable response

### LINKED ISSUE #5021: [Proposal] Use thrift protocol instead of http protocol for master to send heartbeat to other fe to escape unstable heartbeat response
The heartbeat sent by master to other fe is usually read timeout now, 
many log like the following:
2020-12-03 21:21:41,328 WARN 105 [Util.getResultForUrl():361] failed to get result from url: http://xxxxxxxxx:port/api/bootstrap?cluster_id=1xxxxx&token=8456a755-xxxx-4ef9-a460-1547d23fc810. Read timed out
2020-12-03 21:21:41,329 WARN 32 [HeartbeatMgr.runAfterCatalogReady():140] get bad heartbeat response: type: FRONTEND, status: BAD, msg: got exception, name: xxxxxxxxx_9010_1602645254326, queryPort: 0, rpcPort: 0, replayedJournalId: 0
So, I want to use thrift protocol instead of http protocol for master to send heartbeat to other fe to escape unstable heartbeat response, just like broker and be heartbeat

### PR DESCRIPTION
**Pull request by caiconghui**
## Proposed changes

This PR is to support fe master get fe heartbeat response by thrift protocol instead of http protocol.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] Documentation Update (if none of the other choices apply)
- [] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [] I have create an issue on (Fix #5021 ), and have described the bug/feature there in detail
- [] Compiling and unit tests pass locally with my changes
- [] I have added tests that prove my fix is effective or that my feature works
- [] If this change need a document change, I have updated the document
- [] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/common/Config.java [hunk 1]
```diff
@@ -1301,4 +1301,13 @@
      */
     @ConfField
     public static boolean enable_alpha_rowset = false;
+
+    /**
+     * This config is used to solve fe heartbeat response read_timeout problem,
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/org/apache/doris/common/Config.java:1306]: Thrift protocol doesn't have the timeout issue?

Comment (caiconghui) [reply fe/fe-core/src/main/java/org/apache/doris/common/Config.java:1306]: for non master fe, the thread for processing request in fe thrift server is usually idle, and thrift protocol is more lightweight when we keep a long connection opened. But I still confused that why heartbeat by http is easy to readtimout, we just made period http request to fe server, such metric collect,  count running transaction in every db.

Comment (kangkaisen) [reply fe/fe-core/src/main/java/org/apache/doris/common/Config.java:1306]: When HTTP heartbeat read timeout, the FE server has a GC issue?
I think HTTP and thrift protocol are both lightweight, and the heartbeat logic is very simple. So, If the http heartbeat has a timeout issue, there is no reason thrift heartbeat won't have the timeout issue.

Comment (caiconghui) [reply fe/fe-core/src/main/java/org/apache/doris/common/Config.java:1306]: There is no GC issue in fe server. All seems normal, but when non master fe run after sometimes, the read timeout problem occur, and when we restart non master fe, the read timeout issue disappear, and we are not sure that when to meet the read timeout problem again.

Comment (caiconghui) [reply fe/fe-core/src/main/java/org/apache/doris/common/Config.java:1306]: 2020-12-07 14:30:59,424 INFO 115 [RestBaseAction.handleRequest():52] receive http request. url=/api/bootstrap?
2020-12-07 14:31:04,431 INFO 128 [RestBaseAction.handleRequest():52] receive http request. url=/api/bootstrap?
2020-12-07 14:31:09,436 INFO 141 [RestBaseAction.handleRequest():52] receive http request. url=/api/bootstrap
2020-12-07 14:31:14,442 INFO 148 [RestBaseAction.handleRequest():52] receive http request. url=/api/bootstrap
2020-12-07 14:31:19,449 INFO 77 [RestBaseAction.handleRequest():52] receive http request. url=/api/bootstrap?
2020-12-07 14:31:24,455 INFO 87 [RestBaseAction.handleRequest():52] receive http request. url=/api/bootstrap
2020-12-07 14:31:29,461 INFO 94 [RestBaseAction.handleRequest():52] receive http request. url=/api/bootstrap?
2020-12-07 14:31:34,466 INFO 104 [RestBaseAction.handleRequest():52] receive http request. url=/api/bootstrap?
2020-12-07 14:31:39,472 INFO 119 [RestBaseAction.handleRequest():52] receive http request. url=/api/bootstrap?
2020-12-07 14:31:44,477 INFO 129 [RestBaseAction.handleRequest():52] receive http request. url=/api/bootstrap?
2020-12-07 14:31:53,897 INFO 137 [RestBaseAction.handleRequest():52] receive http request. url=/api/bootstrap?
2020-12-07 14:31:56,540 INFO 150 [RestBaseAction.handleRequest():52] receive http request. url=/api/bootstrap?
2020-12-07 14:32:01,545 INFO 79 [RestBaseAction.handleRequest():52] receive http request. url=/api/bootstrap?
2020-12-07 14:32:06,550 INFO 89 [RestBaseAction.handleRequest():52] receive http request. url=/api/bootstrap?
2020-12-07 14:32:11,556 INFO 96 [RestBaseAction.handleRequest():52] receive http request. url=/api/bootstrap

The heartbeat logic is so simple and no blocked logic to process,  but we find receive request has delayed somtimes  according to the log info(every 5 seconds should receive the heartbeat request)

Comment (kangkaisen) [reply fe/fe-core/src/main/java/org/apache/doris/common/Config.java:1306]: I see. Thanks you.

File: gensrc/thrift/FrontendService.thrift [hunk 2]
```diff
@@ -654,6 +654,25 @@ struct TSnapshotLoaderReportRequest {
     5: optional i32 total_num
 }
 
+enum TFrontendPingFrontendStatusCode {
+   OK = 0,
+   FAILED = 1
+}
+
+struct TFrontendPingFrontendRequest {
+   1: required i32 clusterId
```
Comment (sduzh) [inline gensrc/thrift/FrontendService.thrift:663]: `required` field should be used very carefully,  reference [here](https://diwakergupta.github.io/thrift-missing-guide/#_defining_structs)

File: gensrc/thrift/FrontendService.thrift [hunk 3]
```diff
@@ -689,4 +708,6 @@ service FrontendService {
     TStreamLoadPutResult streamLoadPut(1: TStreamLoadPutRequest request)
 
     Status.TStatus snapshotLoaderReport(1: TSnapshotLoaderReportRequest request)
-}
+
+    TBootstrapResult ping(1: TFrontendPingFrontendRequest request)
```
Comment (sduzh) [inline gensrc/thrift/FrontendService.thrift:712]: The request and response names are unmatched with the service name.

File: gensrc/thrift/FrontendService.thrift [hunk 4]
```diff
@@ -654,6 +654,25 @@ struct TSnapshotLoaderReportRequest {
...
+   FAILED = 1
+}
+
+struct TFrontendPingFrontendRequest {
+   1: required i32 clusterId
+   2: required string token
+}
+
+struct TBootstrapResult {
+    1: required TFrontendPingFrontendStatusCode status
```
Comment (sduzh) [inline gensrc/thrift/FrontendService.thrift:668]: Is it ok to use `TStatus` here?

File: fe/fe-core/src/main/java/org/apache/doris/system/HeartbeatMgr.java [hunk 5]
```diff
@@ -282,13 +276,20 @@ public HeartbeatResponse call() {
...
+                            Catalog.getCurrentCatalog().getMaxJournalId(), System.currentTimeMillis(),
                             Version.DORIS_BUILD_VERSION + "-" + Version.DORIS_BUILD_SHORT_HASH);
                 } else {
                     return new FrontendHbResponse(fe.getNodeName(), "not ready");
                 }
             }
+            if (Config.enable_fe_heartbeat_by_thrift) {
+                return getHeartbeatResponseByThrift();
+            } else {
+                return getHeartbeatResponseByHttp();
```
[the remark this answers is not in the document]
Comment (caiconghui) [reply fe/fe-core/src/main/java/org/apache/doris/system/HeartbeatMgr.java:288]: @sduzh be compatible with older versions(0.13), and should be moved next version(0.15). if you ugrade non master fe first, the heartbeat will fail.

File: gensrc/thrift/FrontendService.thrift (hunk 2 above)
Comment (caiconghui) [reply gensrc/thrift/FrontendService.thrift:663]: I think if the filed is really required and the TFrontendPingFrontendRequest is firstly constructed to use, the field used requried is ok

File: gensrc/thrift/FrontendService.thrift (hunk 4 above)
Comment (caiconghui) [reply gensrc/thrift/FrontendService.thrift:668]: TStatus is used for be, just like TBrokerOperationStatus is used for broker, so I think use new status for fe would be better

File: gensrc/thrift/FrontendService.thrift (hunk 3 above)
Comment (caiconghui) [reply gensrc/thrift/FrontendService.thrift:712]: thank you for you advice, I will change it.

File: gensrc/thrift/FrontendService.thrift (hunk 4 above)
Comment (sduzh) [reply gensrc/thrift/FrontendService.thrift:668]: Didn't find any specification that `TStatus` can only be used between BE and FE.
https://github.com/apache/incubator-doris/blob/49f7eb69bfa9190612e84d5cbe989fe15fb4e995/gensrc/thrift/Status.thrift#L22

File: gensrc/thrift/FrontendService.thrift (hunk 3 above)
Comment (sduzh) [reply gensrc/thrift/FrontendService.thrift]: If service is named as `ping`, why not just name the request/response as `TPingRequest`/`TPingResponse`.

Comment (sduzh) [reply gensrc/thrift/FrontendService.thrift]: Sorry to intervene, It's just my personal options, you can just ignore them all ^_^

Comment (caiconghui) [reply gensrc/thrift/FrontendService.thrift]: In my option, by TFrontendPingFrontendRequest, we can know the request is sent by fe to fe, by TPingRequest is less information?

File: gensrc/thrift/FrontendService.thrift (hunk 4 above)
Comment (caiconghui) [reply gensrc/thrift/FrontendService.thrift:668]: Yes, I am also not sure to reuse the TStatus or just like 
```

enum TBrokerOperationStatusCode {
    OK = 0;
    END_OF_FILE = 301;
    
    // user input error
    NOT_AUTHORIZED = 401;
    DUPLICATE_REQUEST = 402;
    INVALID_INPUT_OFFSET = 403; // user input offset is invalid, is large than file length
    INVALID_INPUT_FILE_PATH = 404;
    INVALID_ARGUMENT = 405;
    
    // internal server error
    FILE_NOT_FOUND = 501;
    TARGET_STORAGE_SERVICE_ERROR = 502; // the target storage service error
    OPERATION_NOT_SUPPORTED = 503; // the api is not implemented
}
```
define new status struct

File: gensrc/thrift/FrontendService.thrift (hunk 3 above)
Comment (caiconghui) [reply gensrc/thrift/FrontendService.thrift]: I get what you mean, actually,   TFrontendPingFrontendRequest is like the name in   
```
 TBrokerOperationStatus ping(1: TBrokerPingBrokerRequest request);
```

**Review by kangkaisen — APPROVED** [review]
+1

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**